bluetooth: dis: add Kconfig for disabling optional characteristics

kapi-no opened the pull request:
Added new Kconfig option set that allows the user to control the presence of the following optional characteristics:

- Manufacturer Name String
- Model Number String

Depreacted the old configuration that can be deleted in the future Zephyr releases.

kapi-no committed:
bluetooth: dis: add Kconfig for disabling optional characteristics

Added new Kconfig option set that allows the user to control the
presence of the following optional characteristics:

- Manufacturer Name String
- Model Number String

Depreacted the old configuration that can be deleted in the future
Zephyr releases.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

theob-pro reviewed (changes requested):
Please write about those changes in the release note and the migration guide for 4.1

MarekPieta commented on subsys/bluetooth/services/dis.c:
```diff
@@ -128,12 +151,20 @@ static uint8_t dis_ieee_rcdl[CONFIG_BT_DIS_STR_MAX + 1] = CONFIG_BT_DIS_IEEE_RCD
+	 CONFIG_BT_DIS_SERIAL_NUMBER || CONFIG_BT_DIS_FW_REV || CONFIG_BT_DIS_HW_REV || \
+	 CONFIG_BT_DIS_SW_REV || CONFIG_BT_DIS_IEEE_RCDL)
+
+#if BT_DIS_READ_STR_USED
```
Can't we rely on linker to remove the function if unused? We could add `ARG_UNUSED(read_str);` to prevent build warnings if needed

MarekPieta replied:
Discussed offline. We can leave it as is (finding proper place for `ARG_UNUSED(read_str);` is problematic)

MarekPieta commented on subsys/bluetooth/services/dis.c:
```diff
@@ -32,6 +32,11 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(bt_dis);
 
+BUILD_ASSERT(!IS_ENABLED(CONFIG_BT_DIS_MODEL_NUMBER) ||
+	     !IS_ENABLED(CONFIG_BT_DIS_MODEL_DEPRECATED_USED));
+BUILD_ASSERT(!IS_ENABLED(CONFIG_BT_DIS_MANUF_NAME) ||
+	     !IS_ENABLED(CONFIG_BT_DIS_MANUF_DEPRECATED_USED));
```
Maybe we could build assert only if `CONFIG_BT_DIS_MANUF_NAME_STR` does not match `CONFIG_BT_DIS_MANUF`? (that seems like the only case that is actually problematic for the implementation)

MarekPieta replied:
Alternatively, maybe a Kconfig dependencies could prevent the user from using new options if deprecated ones are still in use

MarekPieta commented on drivers/bluetooth/hci/Kconfig.nxp:
```diff
@@ -48,7 +48,7 @@ endif
 
 if BT_NXP
 
-config BT_DIS_MANUF
```
I would also enable `config BT_DIS_MANUF_NAME` here to ensure proper behavior

MarekPieta commented on samples/bluetooth/peripheral_dis/prj.conf:
```diff
@@ -3,8 +3,8 @@ CONFIG_BT_PERIPHERAL=y
 CONFIG_BT_DIS_PNP=n
-CONFIG_BT_DIS_MODEL="Zephyr Model"
-CONFIG_BT_DIS_MANUF="Zephyr"
+CONFIG_BT_DIS_MODEL_NUMBER_STR="Zephyr Model"
```
I would also enable `CONFIG_BT_DIS_MODEL_NUMBER` and `CONFIG_BT_DIS_MANUF_NAME`

kapi-no committed:
doc: releases: document the DIS changes in changelog and migration guide

Added a changelog entry in the Deprecated and New categories to document
the Kconfig changes for the Bluetooth Device Information Service (DIS).
Updated the migration guide to provide instructions on how to
transition from the old configuration scheme to the new one.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no commented:
> Please write about those changes in the release note and the migration guide for 4.1

Done

jhedberg commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -23,17 +23,57 @@ config BT_DIS_STR_MAX
 
-config BT_DIS_MODEL
-	string "Model name"
+config BT_DIS_MODEL_NUMBER
+	bool "Model number characteristic"
+	depends on !BT_DIS_MODEL_DEPRECATED_USED
+	default y
```
Maybe I missed it in some of the review comments, but why do you need the separate `bool` options alongside of the string options? Can't you just have the string options and have them e.g. `depends on !BT_DIS_MODEL_DEPRECATED_USED`?

kapi-no replied:
The rationale for adding these Kconfig changes is to allow the configuration of the DIS without the Model Number and Manufacturer Name characteristics. The current implementation assumes that these characteristics are always present in the DIS as they cannot be disabled with Kconfigs. However, both characteristics are optional according to the DIS specification. I used the approach that we have for other characteristics that can be enabled/disabled in this service:

New Kconfigs from this PR:
BT_DIS_MODEL_NUMBER <-> BT_DIS_MODEL_NUMBER_STR
BT_DIS_MANUF_NAME <-> BT_DIS_MANUF_NAME_STR
(configures the presence of the given characteristic in the DIS) <-> (configures the string value returned by characteristic)

Existing Kconfigs in the main branch:
BT_DIS_SERIAL_NUMBER <-> BT_DIS_SERIAL_NUMBER_STR
BT_DIS_FW_REV <-> BT_DIS_FW_REV_STR
BT_DIS_HW_REV <-> BT_DIS_HW_REV_STR
(...)
(configures the presence of the given characteristic in the DIS) <-> (configures the string value returned by characteristic)

jhedberg replied:
The argument for consistency with existing options makes sense, but I don't see what this has to do with the optional nature of these features, since you can just check for empty strings to decide to include support (e.g. something like `sizeof(CONFIG_BT_DIS_MANUF_NAME_STR)` should be evaluatable at build time)
